File: esphome/components/graphical_layout/display_rendering_panel.py
import esphome.codegen as cg
from esphome.components.display import DisplayRef, display_ns
import esphome.config_validation as cv
from esphome.const import CONF_HEIGHT, CONF_LAMBDA, CONF_WIDTH

graphical_layout_ns = cg.esphome_ns.namespace("graphical_layout")
Rect = display_ns.class_("Rect")
DisplayRenderingPanel = graphical_layout_ns.class_("DisplayRenderingPanel")

CONF_DISPLAY_RENDERING_PANEL = "display_rendering_panel"


def get_config_schema(base_item_schema, item_type_schema):
    return base_item_schema.extend(
        {
            cv.GenerateID(): cv.declare_id(DisplayRenderingPanel),
            cv.Required(CONF_WIDTH): cv.templatable(cv.int_range(min=1)),
            cv.Required(CONF_HEIGHT): cv.templatable(cv.int_range(min=1)),
            cv.Required(CONF_LAMBDA): cv.lambda_,
        }
    )


async def config_to_layout_item(pvariable_builder, item_config, child_item_builder):
    var = await pvariable_builder(item_config)

    width = await cg.templatable(item_config[CONF_WIDTH], args=[], output_type=cg.int_)
    cg.add(var.set_width(width))

    height = await cg.templatable(
        item_config[CONF_HEIGHT], args=[], output_type=cg.int_
    )
    cg.add(var.set_height(height))

    lambda_ = await cg.process_lambda(
        item_config[CONF_LAMBDA],
        [(DisplayRef, "it"), (Rect, "bounds")],
        return_type=cg.void,
    )
    cg.add(var.set_lambda(lambda_))

    return var

File: esphome/components/graphical_layout/display_rendering_panel.cpp
#include "display_rendering_panel.h"

#include "esphome/components/display/display.h"
#include "esphome/components/display/rect.h"
#include "esphome/core/log.h"

namespace esphome {
namespace graphical_layout {

static const char *const TAG = "displayrenderingpanel";

void DisplayRenderingPanel::dump_config(int indent_depth, int additional_level_depth) {
  this->dump_config_base_properties(TAG, indent_depth);
  ESP_LOGCONFIG(TAG, "%*sDimensions: %ix%i", indent_depth, "", this->width_.value(), this->height_.value());
  ESP_LOGCONFIG(TAG, "%*sHas drawing lambda: %s", indent_depth, "", YESNO(this->lambda_ != nullptr));
}

display::Rect DisplayRenderingPanel::measure_item_internal(display::Display *display) {
  return display::Rect(0, 0, this->width_.value(), this->height_.value());
}

void DisplayRenderingPanel::render_internal(display::Display *display, display::Rect bounds) {
  this->lambda_(*display, bounds);
}

}  // namespace graphical_layout
}  // namespace esphome
